### PR TITLE
Introduce active cell index 2

### PR DESCRIPTION
**Pull request by bangerth**
This is the first part of the promised follow-up to #765 (active_cell_index) that tries to address #761 . These are some mechanical changes that remove the double indexing loops over cells and indices.

I've excluded the interesting places in DataOut and KellyErrorEstimator where I think that the active cell index will lead to significant code cleanups, and to the steps. Rather, this PR only contains the controversial pieces in that (i) they are in the library where clarity is good but not the overriding principle, (ii) it will lead to some slowdown, (iii) the cleanups are minor.

I think that in all cases treated here, the slowdown will be relatively minor (accessing active_cell_index is relatively cheap compared with most of the surrounding code in most cases). I've kept things in separate commits to make it simple to drop any one people don't like.

In summary, I think these cases lead to small cleanups but I'm otherwise not wedded to them. Please discuss!

### CODE REVIEW

**Comment by kronbichler** [comment]
I don't think there will be a substantial slowdown and I am not at all worried about that in the places where this is used. Loops over cells are not 'innermost' loops. :-)

In general, I support this patch and have no objections. I will wait for others to comment.

**Comment by kronbichler** [comment]
To be clear: I think that the slowdown is really negligible: All you do is a function call (which could be moved to the .templates.h file to avoid that) and two vector accesses by present_level and present_index. Certainly a little bit of pointer arithmetics, but less work than incrementing the active_cell_iterator for example. The memory access itself should be perfectly predictable so nothing to worry about either.

File: source/grid/grid_out.cc [hunk 1]
```diff
@@ -1066,10 +1064,11 @@ void GridOut::write_msh (const Triangulation<dim, spacedim> &tria,
 
   // write faces and lines with
   // non-zero boundary indicator
+  unsigned int next_index = tria.n_active_cells()+1;
   if (msh_flags.write_faces)
-    write_msh_faces (tria, cell_index, out);
+    write_msh_faces (tria, next_index, out);
   if (msh_flags.write_lines)
-    write_msh_lines (tria, cell_index, out);
+    write_msh_lines (tria, next_index, out);
```
Comment (tjhei) [inline source/grid/grid_out.cc]: I am confused what happens here (also in the original code). We give unique ids to cells and faces because we keep counting upwards with next_index, but faces and lines would share the same index? I wonder if writing faces and lines will produce a broken file. Or do we never output faces and lines at the same time? Either way, I think write_msh_faces should return/modify the new next_index.

Btw., it took me a while to see what is going on (continue counting after cells). I think the old code was more clear in that sense. Not a big issue of course.

File: source/grid/grid_out.cc [hunk 2]
```diff
@@ -1191,10 +1188,11 @@ void GridOut::write_ucd (const Triangulation<dim,spacedim> &tria,
 
   // write faces and lines with
   // non-zero boundary indicator
+  unsigned int next_index = tria.n_active_cells()+1;
   if (ucd_flags.write_faces)
-    write_ucd_faces (tria, cell_index, out);
+    write_ucd_faces (tria, next_index, out);
   if (ucd_flags.write_lines)
-    write_ucd_lines (tria, cell_index, out);
+    write_ucd_lines (tria, next_index, out);
```
Comment (tjhei) [inline source/grid/grid_out.cc]: same here.

**Comment by tjhei** [comment]
Looks good (except my one comment) and yes, I don't think looking up the index will make a noticeable difference.

File: source/grid/grid_out.cc (hunk 1 above)
Comment (bangerth) [inline source/grid/grid_out.cc]: Yes, I think this is broken. Will need to look into it. Same below.

**Comment by bangerth** [comment]
I've opened #814 for this issue. It is pre-existing, so would like to address it in a separate PR if you allow me to, rather than conjoining it with the current issue under consideration here.

**Comment by tjhei** [comment]
Agreed. So do we merge this PR as is now? If yes, please do so.
